csharp_dotnetcore/09.Message-Routing/Dialogs/Greeting/GreetingDialog.cs

```diff
@@ -0,0 +1,224 @@
+                                                CancellationToken cancellationToken)
+        {
+            var greetingState = await GreetingStateAccessor.GetAsync(stepContext.Context);
+
+            // if we have everything we need, greet user and return.
+            if (greetingState != null && !string.IsNullOrWhiteSpace(greetingState.Name) && !string.IsNullOrWhiteSpace(greetingState.City))
+            {
+                return await GreetUser(stepContext);
+            }
+
+            if (string.IsNullOrWhiteSpace(greetingState.Name))
+            {
+                // prompt for name, if missing
+                var opts = new PromptOptions
+                {
+                    Prompt = new Activity
+                    {
+                        Type = ActivityTypes.Message,
+                        Text = "What is your name?",
+                    },
+                };
+                return await stepContext.PromptAsync(NamePrompt, opts);
+            }
+            else
+            {
+                return await stepContext.NextAsync();
+            }
+        }
+
+        private async Task<DialogTurnResult> PromptForCityStepAsync(
+                                                        WaterfallStepContext stepContext,
+                                                        CancellationToken cancellationToken)
+        {
+            // Save name, if prompted.
+            var greetingState = await GreetingStateAccessor.GetAsync(stepContext.Context);
+            var lowerCaseName = stepContext.Result as string;
+            if (string.IsNullOrWhiteSpace(greetingState.Name) && lowerCaseName != null)
+            {
+                // Capitalize and set name.
+                greetingState.Name = char.ToUpper(lowerCaseName[0]) + lowerCaseName.Substring(1);
+                await GreetingStateAccessor.SetAsync(stepContext.Context, greetingState);
+            }
+
+            if (string.IsNullOrWhiteSpace(greetingState.City))
+            {
+                var opts = new PromptOptions
+                {
+                    Prompt = new Activity
+                    {
+                        Type = ActivityTypes.Message,
+                        Text = $"Hello {greetingState.Name}, what city do you live in?",
+                    },
+                };
+                return await stepContext.PromptAsync(CityPrompt, opts);
+            }
+            else
+            {
+                return await stepContext.NextAsync();
+            }
+        }
+
+        private async Task<DialogTurnResult> DisplayGreetingStateStepAsync(
+                                                    WaterfallStepContext stepContext,
+                                                    CancellationToken cancellationToken)
+        {
+            // Save city, if prompted.
+            var greetingState = await GreetingStateAccessor.GetAsync(stepContext.Context);
```

Comment:
We should switch greetingState to userProfile and greetingStateAccessor to userProfileAccessor